x/precisebank/types/fractional_balances_test.go

```diff
@@ -0,0 +1,79 @@
+package types_test
+
+import (
+	"testing"
+
+	sdkmath "cosmossdk.io/math"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/kava-labs/kava/app"
+	"github.com/kava-labs/kava/x/precisebank/types"
+	"github.com/stretchr/testify/require"
+)
+
+func TestFractionalBalances_Validate(t *testing.T) {
+	app.SetSDKConfig()
+
+	tests := []struct {
+		name    string
+		fbs     types.FractionalBalances
+		wantErr string
+	}{
+		{
+			"valid - empty",
+			types.FractionalBalances{},
+			"",
+		},
+		{
+			"valid - nil",
+			nil,
+			"",
+		},
+		{
+			"valid - multiple balances",
+			types.FractionalBalances{
+				types.NewFractionalBalance(sdk.AccAddress{1}.String(), sdkmath.NewInt(100)),
+				types.NewFractionalBalance(sdk.AccAddress{2}.String(), sdkmath.NewInt(100)),
+				types.NewFractionalBalance(sdk.AccAddress{3}.String(), sdkmath.NewInt(100)),
+			},
+			"",
+		},
+		{
+			"invalid - single invalid balance",
+			types.FractionalBalances{
+				types.NewFractionalBalance(sdk.AccAddress{1}.String(), sdkmath.NewInt(100)),
+				types.NewFractionalBalance(sdk.AccAddress{2}.String(), sdkmath.NewInt(-1)),
+				types.NewFractionalBalance(sdk.AccAddress{3}.String(), sdkmath.NewInt(100)),
+			},
+			"invalid fractional balance for kava1qg7c45n6: negative amount: -1",
+		},
+		{
+			"invalid - duplicate address",
+			types.FractionalBalances{
+				types.NewFractionalBalance(sdk.AccAddress{1}.String(), sdkmath.NewInt(100)),
+				types.NewFractionalBalance(sdk.AccAddress{1}.String(), sdkmath.NewInt(100)),
+			},
+			"duplicate address: kava1qy0xn7za",
+		},
+		{
+			"invalid - duplicate address with whitespace",
+			types.FractionalBalances{
+				types.NewFractionalBalance(sdk.AccAddress{1}.String(), sdkmath.NewInt(100)),
+				types.NewFractionalBalance(" "+sdk.AccAddress{1}.String(), sdkmath.NewInt(100)),
```

Comment:
Is whitespace in an address valid? Should we re-order checking if a fractional balance is valid so we can ensure that our uniqueness check only happens on a valid collection of FractionalBalance's?

Reply:
ah right, I misunderstood the TrimSpace in `AccAddressFromBech32()`, whitespace in addresses is not valid